pkg/rm/remoting_parser_api.go

```diff
@@ -0,0 +1,116 @@
+package rm
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/seata/seata-go/pkg/common/log"
+)
+
+const (
+	RemotingUnknow  = 0
+	RemotingGrpc    = 1
+	RemotingDubbogo = 2
+	RemotingLocal   = 3
+)
+
+type RemotingParser interface {
+	IsService(target interface{}) bool
+	IsReference(target interface{}) bool
+	IsRemoting(target interface{}) bool
+	ParseService(target interface{}) (*TwoPhaseAction, error)
+	ParseReference(target interface{}) (*TwoPhaseAction, error)
+	ParseTwoPhaseActionByInterface(target interface{}) (*TwoPhaseAction, error)
+	GetRemotingType(target interface{}) int
+}
+
+var (
+	remotingParseTable = make([]RemotingParser, 0)
+)
+
+func RegisterRemotingParse(parse RemotingParser) {
+	remotingParseTable = append(remotingParseTable, parse)
+}
+
+type DefaultRemotingParser struct {
+}
+
+var (
+	once                           = sync.Once{}
+	defaultRemotingParserSingleton *DefaultRemotingParser
+)
+
+func GetDefaultRemotingParser() *DefaultRemotingParser {
+	if defaultRemotingParserSingleton == nil {
+		once.Do(func() {
+			defaultRemotingParserSingleton = &DefaultRemotingParser{}
+		})
+	}
+	return defaultRemotingParserSingleton
+
+}
+
+func (d *DefaultRemotingParser) IsService(target interface{}) bool {
+	for _, v := range remotingParseTable {
+		if v.IsService(target) {
+			return true
+		}
+	}
+	return false
+}
+func (d *DefaultRemotingParser) IsReference(target interface{}) bool {
+	for _, v := range remotingParseTable {
+		if v.IsReference(target) {
+			return true
+		}
+	}
+	return false
+}
+func (d *DefaultRemotingParser) IsRemoting(target interface{}) bool {
+	for _, v := range remotingParseTable {
+		if v.IsRemoting(target) {
+			return true
+		}
+	}
+	return false
+}
+func (d *DefaultRemotingParser) ParseService(target interface{}) (*TwoPhaseAction, error) {
+	for _, v := range remotingParseTable {
+		if res, err := v.ParseService(target); err != nil {
+			log.Warnf("parse service failed : %s", err)
+		} else if res != nil {
+			return res, nil
+		}
+	}
+	return nil, errors.New(fmt.Sprintf("don't find a parser to resolve service %v", target))
```

Comment:
```go
for (xxxx) {
   res, err := v.ParseService(target); 
   if err != nil || res == nil{
        log.Warnf("parse service failed : %s", err)
        continue
  }
  return res, nil 
}
return nil ,err
```
may be this will be better

Reply:
err != nill  is the wrong case, but res = = nil is the normal case, no warn need to be logged

Reply:
要不就不打印 `log.Warnf("parse service failed : %s", err)`，其实打印了，意义也不大。主要是我觉得 if else 太丑了。